### DISCUSSION TITLE
Building MD document from multiple files / remark include import partial

### DISCUSSION THREAD

**Discussion by slavafomin** [remarkjs/remark#1101] [Q&A] [answered]
Hello!

Thank you for your great work, remark looks fantastic!

However, I'm trying to do a pretty basic thing — I have several MD-files that I want to integrate together into a single final markdown file, but I can't figure out how to do this. All the include/import plugins available are severely outdated and doesn't work with the latest version of remark.

Could you recommend a solution for my use case please? Thank you!

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @slavafomin! 👋 

> I have several MD-files that I want to integrate together into a single final markdown file

Can you expand a bit on what you mean here?
Specifically:
* What is the expected output format? (Markdown, HTML, React Component)
* How do the files need to be integrated? (Just concatenated together, something more template like)
* Who is going to be authoring the documents? (Developers, Technical writers, etc)

Without knowing that context I can only offer some broad suggestions:

### If you want something akin to a template language

Consider MDX https://mdxjs.com/
Which builds on `remark`, adding JSX into Markdown, which allows for imports, exports, components, and embedded JavaScript (which can be used for templates).

### If you want documents embedded in specific spots in other documents only

Consider using https://github.com/remarkjs/remark-directive then creating a new plugin https://unifiedjs.com/learn/ which replaces the directive with the document.

### If you want to concatenate the documents together, leaving the content unchanged

You may not need `remark`, plain ol' unix/linux/MacOs [`cat`](https://en.wikipedia.org/wiki/Cat_(Unix)) may be a better fit.

**Reply by slavafomin** [reply]
Hello @ChristianMurphy! Thank you for a quick and elaborate answer. I will try to explain my case further. I'm working on a library and I want to generate a README.md file for it to be displayed by e.g. GitHub, npm, etc. So it's MD++ → MD transformation. Ideally, I would like to have an `index.md` file that have something like `<!-- @include: license.md -->` in it. This will help to make things more manageable when you have a very big README. Also, I would like to embed some content from other sources, e.g. results of benchmarks, code examples etc.

Regarding your suggestions:

- If I understand correctly the MDX will output HTML, while I want to output MD that can be read directly if viewer is not available.
- The directive approach looks like a good one, but I'm a little unsettled here. The include/embed functionality is a pretty basic stuff. it looks weird that I have to write a plugin for it myself considering that remark is a mature and established solution. This leads me to a thought that maybe I'm using it wrong, i.e. bad tool for the job?

So, maybe I should consider another approach, e.g. to use some templating system like nunjucks to build the markdown from multiple files and then use remark for specific things like TOC and such?

**Reply by ChristianMurphy** [reply]
> I would like to embed some content from other sources, e.g. results of benchmarks, code examples etc.

https://www.npmjs.com/package/remark-code-import

> If I understand correctly the MDX will output HTML, while I want to output MD that can be read directly if viewer is not available.

MDX can also serialize back to MDX.

> The include/embed functionality is a pretty basic stuff.

Not really.
None of HTML, Markdown, or AsciiDoc have this.
They're static document languages not template languages.

> This leads me to a thought that maybe I'm using it wrong, i.e. bad tool for the job?

If you want to render Markdown on GitHub and NPM.
It's a good choice.

> maybe I should consider another approach, e.g. to use some templating system like nunjucks to build the markdown from multiple files

It can be problematic, see https://github.com/orgs/remarkjs/discussions/1094 for a recent discussions on pre-processing templates and the challenges associated.

**Reply by slavafomin** [reply]
Thanks for the reply!

> MDX can also serialize back to MDX.

Well, I don't want to serialize MDX to MDX (is it a no-op?). Is it possible to serialize MDX to MD though? For example to output a generated markdown list? Like this:

```md
- foo
- bar
- baz
```

> None of HTML, Markdown, or AsciiDoc have this.
> They're static document languages not template languages.

That's why I'm looking into Markdown processors like Remark. And for such processor "includes" should be a pretty basic feature, doesn't it?

> It can be problematic, see … for a recent discussions on pre-processing templates and the challenges associated.

I've actually tried to use Nunjucks with Remark and so far it works pretty good. Nunjucks could do all kind of custom content generation (like includes), while Remark plugins could do some MD-specific stuff like TOCs, etc.

> https://www.npmjs.com/package/remark-code-import

See, this is weird. You have plugin to import code, but no plugin to import another MD :)
I still can't figure out why.

**Reply by ChristianMurphy** [reply]
> I don't want to serialize MDX to MDX (is it a no-op?). 

It is not a no-op, it can do transformation and formatting in the process.
Including replacing something like `<Embed href="https://example.com/content" />` with plain Markdown from that `href` location.

> And for such processor "includes" should be a pretty basic feature, doesn't it?

AST's make adding, removing, and otherwise modifying content easier.
Whether than makes includes "basic" is another question.
Ensuring the includes have easy to understand syntax, are secure, are robust, and are performant can be more nuanced and complex.
Given the lack of specifics and interest in developing it, I suspect [the devil is in the details](https://en.wikipedia.org/wiki/The_devil_is_in_the_details).

> I've actually tried to use Nunjucks with Remark and so far it works pretty good. Nunjucks could do all kind of custom content generation (like includes)

If it works for you, go for it!
The conversation in https://github.com/orgs/remarkjs/discussions/1094 revolves around security constraints (you need to trust the source if you are injecting raw content), and performance (two phase builds tend to be slower), if you don't have those constraints, then you don't need to worry about them.

> See, this is weird. You have plugin to import code, but no plugin to import another MD :)
> I still can't figure out why.

Many (if not most) plugins are community contributed, including `remark-code-import`.
There isn't one for your particular use case because no one has been interested and motivated enough to make one, yet.

**Reply by ChristianMurphy** [reply]
If you're interested in creating a new plugin.
There are some guides to the unified ecosystem and plugin development at https://unifiedjs.com/learn/

**Reply by slavafomin** [reply]
> Including replacing something like `<Embed href="https://example.com/content" />` with plain Markdown from that `href` location.

Thank you for clarifying that. By the way, do you have an example of the MDX code that can generate MD and not HTML? If I can use MDX to generate MD than this would be a better approach than using two-step processing.

> There isn't one for your particular use case because no one has been interested and motivated enough to make one, yet.

That's what baffle me a lot actually :)

**Reply by ChristianMurphy** [reply]
> By the way, do you have an example of the MDX code that can generate MD and not HTML?

https://mdxjs.com/packages/remark-mdx/
